File: browser/ai_browser_agent/lib/tools.mjs
import fs from 'node:fs/promises';
import path from 'node:path';

import {
  activateTarget,
  ChromeSession,
  closeTarget,
  createTarget,
  getChromeVersion,
  launchChrome,
  listTargets,
} from './chrome_client.mjs';
import {
  AiManifestBundleDefinition,
  AiToolDanger,
  AiToolDefinition,
  clampNumber,
  schemaJson,
} from './models.mjs';

export function defaultBundles() {
  return [
    new AiManifestBundleDefinition({
      id: 'service',
      description: 'Adapter health, logs, config, and recent calls.',
      prefixes: ['service.'],
    }),
    new AiManifestBundleDefinition({
      id: 'browser.targets',
      description: 'Chrome connectivity, target discovery, and page lifecycle tools.',
      prefixes: ['browser.chrome_', 'browser.targets_', 'browser.page_open', 'browser.page_activate', 'browser.page_close', 'browser.current_target_'],
    }),
    new AiManifestBundleDefinition({
      id: 'browser.inspect',
      description: 'Runtime evaluation, DOM inspection, and screenshot capture.',
      prefixes: ['browser.page_navigate', 'browser.page_reload', 'browser.runtime_', 'browser.dom_', 'browser.page_capture_'],
    }),
  ];
}

export function registerDefaultTools(registry, context) {
  const register = (definition) => registry.register(new AiToolDefinition(definition));

  register({
    id: 'service.health_get',
    description: 'Return the adapter health payload.',
    argsSchemaJson: schemaJson({ type: 'object', additionalProperties: false, properties: {} }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'service.health_get',
    handler: async () => context.buildHealthPayload({ includeOk: true }),
  });

  register({
    id: 'service.agent_brief_get',
    description: 'Return the concise operating brief for this adapter.',
    argsSchemaJson: schemaJson({ type: 'object', additionalProperties: false, properties: {} }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'service.agent_brief_get',
    handler: async () => context.buildAgentBriefPayload({ includeOk: true }),
  });

  register({
    id: 'service.config_get',
    description: 'Return effective adapter configuration, token header names, and the current target state.',
    argsSchemaJson: schemaJson({ type: 'object', additionalProperties: false, properties: {} }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'service.config_get',
    handler: async () => ({
      rootDir: context.config.rootDir,
      requireToken: context.config.requireToken,
      fullAccessEnabled: context.config.fullAccessEnabled,
      acceptedTokenHeaders: context.config.acceptedTokenHeaders,
      chromeBaseUrl: context.config.chromeBaseUrl,
      currentTargetId: context.browserState.currentTargetId || null,
      lastLaunchInfo: context.browserState.lastLaunchInfo,
    }),
  });

  register({
    id: 'service.logs_get',
    description: 'List recent service logs.',
    argsSchemaJson: schemaJson({
      type: 'object',
      additionalProperties: false,
      properties: {
        maxEntries: { type: 'integer', minimum: 1, maximum: 300 },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'service.logs_get',
    handler: async (args) => context.resultHandleStore.buildItemsResult({
      sourceToolId: 'service.logs_get',
      fieldName: 'logs',
      items: context.runtimeState.recentLogs(args.maxEntries),
      summary: {
        kind: 'logs',
      },
      pageSize: clampNumber(Number(args.maxEntries) || 100, 1, 200),
    }),
  });

  register({
    id: 'service.recent_tool_calls_get',
    description: 'List recent tool call outcomes recorded by the adapter.',
    argsSchemaJson: schemaJson({
      type: 'object',
      additionalProperties: false,
      properties: {
        maxEntries: { type: 'integer', minimum: 1, maximum: 300 },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'service.recent_tool_calls_get',
    handler: async (args) => context.resultHandleStore.buildItemsResult({
      sourceToolId: 'service.recent_tool_calls_get',
      fieldName: 'calls',
      items: context.runtimeState.recentCalls(args.maxEntries),
      summary: {
        kind: 'toolCalls',
      },
      pageSize: clampNumber(Number(args.maxEntries) || 100, 1, 200),
    }),
  });

  register({
    id: 'service.token_regenerate',
    description: 'Regenerate the adapter token and return the new token value.',
    argsSchemaJson: schemaJson({ type: 'object', additionalProperties: false, properties: {} }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'service.token_regenerate',
    danger: AiToolDanger.medium,
    requiresConfirmation: true,
    handler: async () => ({
      token: await context.regenerateToken(),
      acceptedTokenHeaders: context.config.acceptedTokenHeaders,
    }),
  });

  register({
    id: 'browser.chrome_status_get',
    description: 'Return Chrome remote debugging availability, version metadata, and target counts.',
    argsSchemaJson: schemaJson({ type: 'object', additionalProperties: false, properties: {} }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.chrome_status_get',
    handler: async () => {
      try {
        const version = await getChromeVersion(context.config);
        const targets = await listTargets(context.config);
        return {
          reachable: true,
          version,
          targetCount: targets.length,
          currentTargetId: context.browserState.currentTargetId || null,
        };
      } catch (error) {
        return {
          reachable: false,
          error: error.message,
          currentTargetId: context.browserState.currentTargetId || null,
        };
      }
    },
  });

  register({
    id: 'browser.chrome_launch',
    description: 'Launch Chrome with remote debugging enabled for this adapter.',
    argsSchemaJson: schemaJson({
      type: 'object',
      additionalProperties: false,
      properties: {
        url: { type: 'string' },
        headless: { type: 'boolean' },
        userDataDir: { type: 'string' },
        timeoutMs: { type: 'integer', minimum: 1000, maximum: 60000 },
        extraArgs: { type: 'array', items: { type: 'string' } },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.chrome_launch',
    danger: AiToolDanger.medium,
    handler: async (args) => {
      const launchInfo = await launchChrome(context.config, {
        url: args.url ? String(args.url) : '',
        headless: Boolean(args.headless),
        userDataDir: args.userDataDir ? resolveWithinRoot(context.config.rootDir, String(args.userDataDir)) : '',
        timeoutMs: args.timeoutMs,
        extraArgs: Array.isArray(args.extraArgs) ? args.extraArgs : [],
      });
      context.browserState.lastLaunchInfo = launchInfo;
      return launchInfo;
    },
  });

  register({
    id: 'browser.targets_list',
    description: 'List remote debugging targets known to Chrome.',
    argsSchemaJson: schemaJson({
      type: 'object',
      additionalProperties: false,
      properties: {
        type: { type: 'string' },
        limit: { type: 'integer', minimum: 1, maximum: 200 },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.targets_list',
    handler: async (args) => {
      let targets = await listTargets(context.config);
      if (args.type) {
        targets = targets.filter((target) => target.type === String(args.type));
      }
      const normalized = targets.map(normalizeTarget);
      if (!context.browserState.currentTargetId && normalized.length > 0) {
        context.browserState.currentTargetId =
          normalized.find((target) => target.type === 'page')?.id ?? normalized[0].id;
      }
      return context.resultHandleStore.buildItemsResult({
        sourceToolId: 'browser.targets_list',
        fieldName: 'targets',
        items: normalized,
        summary: {
          type: args.type || '',
        },
        pageSize: clampNumber(Number(args.limit) || 20, 1, 200),
      });
    },
  });

  register({
    id: 'browser.current_target_get',
    description: 'Return the adapter current target, resolving a default page target if necessary.',
    argsSchemaJson: schemaJson({ type: 'object', additionalProperties: false, properties: {} }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.current_target_get',
    handler: async () => {
      const target = await resolveTarget(context, '');
      return {
        currentTargetId: context.browserState.currentTargetId,
        target: normalizeTarget(target),
      };
    },
  });

  register({
    id: 'browser.page_open',
    description: 'Open a new page target in Chrome and make it the current target.',
    argsSchemaJson: schemaJson({
      type: 'object',
      required: ['url'],
      additionalProperties: false,
      properties: {
        url: { type: 'string' },
        activate: { type: 'boolean' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.page_open',
    handler: async (args) => {
      const created = await createTarget(context.config, String(args.url));
      context.browserState.currentTargetId = created.id;
      if (args.activate !== false) {
        await activateTarget(context.config, created.id);
      }
      return {
        target: normalizeTarget(created),
        currentTargetId: context.browserState.currentTargetId,
      };
    },
  });

  register({
    id: 'browser.page_activate',
    description: 'Activate an existing target and make it the current target.',
    argsSchemaJson: schemaJson({
      type: 'object',
      required: ['targetId'],
      additionalProperties: false,
      properties: {
        targetId: { type: 'string' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.page_activate',
    handler: async (args) => {
      const targetId = String(args.targetId);
      const response = await activateTarget(context.config, targetId);
      context.browserState.currentTargetId = targetId;
      return {
        targetId,
        status: response.trim(),
      };
    },
  });

  register({
    id: 'browser.page_close',
    description: 'Close an existing target and clear it if it was the current target.',
    argsSchemaJson: schemaJson({
      type: 'object',
      required: ['targetId'],
      additionalProperties: false,
      properties: {
        targetId: { type: 'string' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.page_close',
    danger: AiToolDanger.medium,
    handler: async (args) => {
      const targetId = String(args.targetId);
      const response = await closeTarget(context.config, targetId);
      if (context.browserState.currentTargetId === targetId) {
        context.browserState.currentTargetId = '';
      }
      return {
        targetId,
        status: response.trim(),
        currentTargetId: context.browserState.currentTargetId || null,
      };
    },
  });

  register({
    id: 'browser.page_navigate',
    description: 'Navigate a page target to a new URL through Chrome DevTools Protocol.',
    argsSchemaJson: schemaJson({
      type: 'object',
      required: ['url'],
      additionalProperties: false,
      properties: {
        url: { type: 'string' },
        targetId: { type: 'string' },
        waitForLoad: { type: 'boolean' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.page_navigate',
    handler: async (args) => {
      const target = await resolveTarget(context, args.targetId ? String(args.targetId) : '');
      return withTargetSession(context, target, async (session) => {
        await session.send('Page.enable');
        const shouldWait = args.waitForLoad !== false;
        const waitForLoad = shouldWait ? session.waitForEvent('Page.loadEventFired', { timeoutMs: 15000 }) : null;
        const navigation = await session.send('Page.navigate', { url: String(args.url) });
        if (waitForLoad) {
          try {
            await waitForLoad;
          } catch {
            // Navigation may still succeed even if the load event is delayed.
          }
        }
        return {
          target: normalizeTarget(target),
          navigation,
          waitedForLoad: shouldWait,
        };
      });
    },
  });

  register({
    id: 'browser.page_reload',
    description: 'Reload a page target through Chrome DevTools Protocol.',
    argsSchemaJson: schemaJson({
      type: 'object',
      additionalProperties: false,
      properties: {
        targetId: { type: 'string' },
        ignoreCache: { type: 'boolean' },
        waitForLoad: { type: 'boolean' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.page_reload',
    handler: async (args) => {
      const target = await resolveTarget(context, args.targetId ? String(args.targetId) : '');
      return withTargetSession(context, target, async (session) => {
        await session.send('Page.enable');
        const shouldWait = args.waitForLoad !== false;
        const waitForLoad = shouldWait ? session.waitForEvent('Page.loadEventFired', { timeoutMs: 15000 }) : null;
        await session.send('Page.reload', { ignoreCache: Boolean(args.ignoreCache) });
        if (waitForLoad) {
          try {
            await waitForLoad;
          } catch {
            // ignore
          }
        }
        return {
          target: normalizeTarget(target),
          reloaded: true,
          waitedForLoad: shouldWait,
        };
      });
    },
  });

  register({
    id: 'browser.runtime_evaluate',
    description: 'Evaluate JavaScript in the selected page target and return the DevTools Protocol result.',
    argsSchemaJson: schemaJson({
      type: 'object',
      required: ['expression'],
      additionalProperties: false,
      properties: {
        expression: { type: 'string' },
        targetId: { type: 'string' },
        awaitPromise: { type: 'boolean' },
        returnByValue: { type: 'boolean' },
        includeCommandLineAPI: { type: 'boolean' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.runtime_evaluate',
    handler: async (args) => {
      const target = await resolveTarget(context, args.targetId ? String(args.targetId) : '');
      return withTargetSession(context, target, async (session) => {
        await session.send('Runtime.enable');
        const evaluation = await session.send('Runtime.evaluate', {
          expression: String(args.expression),
          awaitPromise: args.awaitPromise !== false,
          returnByValue: args.returnByValue !== false,
          includeCommandLineAPI: Boolean(args.includeCommandLineAPI),
        });
        return {
          target: normalizeTarget(target),
          evaluation,
        };
      });
    },
  });

  register({
    id: 'browser.dom_document_get',
    description: 'Read the DOM document tree for the selected page target.',
    argsSchemaJson: schemaJson({
      type: 'object',
      additionalProperties: false,
      properties: {
        targetId: { type: 'string' },
        depth: { type: 'integer', minimum: -1, maximum: 16 },
        pierce: { type: 'boolean' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.dom_document_get',
    handler: async (args) => {
      const target = await resolveTarget(context, args.targetId ? String(args.targetId) : '');
      return withTargetSession(context, target, async (session) => {
        await session.send('DOM.enable');
        const documentResult = await session.send('DOM.getDocument', {
          depth: Number.isInteger(args.depth) ? args.depth : 4,
          pierce: Boolean(args.pierce),
        });
        return context.resultHandleStore.buildTextResult({
          sourceToolId: 'browser.dom_document_get',
          text: JSON.stringify({
            target: normalizeTarget(target),
            document: documentResult.root,
          }, null, 2),
          summary: {
            targetId: target.id,
          },
        });
      });
    },
  });

  register({
    id: 'browser.dom_query_selector_all',
    description: 'Query the DOM by CSS selector and return matching node summaries.',
    argsSchemaJson: schemaJson({
      type: 'object',
      required: ['selector'],
      additionalProperties: false,
      properties: {
        selector: { type: 'string' },
        targetId: { type: 'string' },
        limit: { type: 'integer', minimum: 1, maximum: 100 },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.dom_query_selector_all',
    handler: async (args) => {
      const target = await resolveTarget(context, args.targetId ? String(args.targetId) : '');
      return withTargetSession(context, target, async (session) => {
        await session.send('DOM.enable');
        const documentResult = await session.send('DOM.getDocument', { depth: 1, pierce: true });
        const matches = await session.send('DOM.querySelectorAll', {
          nodeId: documentResult.root.nodeId,
          selector: String(args.selector),
        });
        const safeLimit = clampNumber(Number(args.limit) || 20, 1, 100);
        const items = [];
        for (const nodeId of (matches.nodeIds ?? []).slice(0, safeLimit)) {
          const node = await session.send('DOM.describeNode', { nodeId, depth: 0 });
          items.push(normalizeDomNode(node.node));
        }
        return context.resultHandleStore.buildItemsResult({
          sourceToolId: 'browser.dom_query_selector_all',
          fieldName: 'nodes',
          items,
          summary: {
            selector: String(args.selector),
            targetId: target.id,
            matchCount: (matches.nodeIds ?? []).length,
          },
          pageSize: safeLimit,
        });
      });
    },
  });

  register({
    id: 'browser.page_capture_screenshot',
    description: 'Capture a screenshot from the selected page target and write it to disk.',
    argsSchemaJson: schemaJson({
      type: 'object',
      required: ['outputPath'],
      additionalProperties: false,
      properties: {
        outputPath: { type: 'string' },
        targetId: { type: 'string' },
        format: { type: 'string', enum: ['png', 'jpeg', 'webp'] },
        quality: { type: 'integer', minimum: 0, maximum: 100 },
        fullPage: { type: 'boolean' },
        fromSurface: { type: 'boolean' },
      },
    }),
    returnSchemaJson: schemaJson({ type: 'object' }),
    handlerName: 'browser.page_capture_screenshot',
    danger: AiToolDanger.high,
    requiresConfirmation: true,
    handler: async (args) => {
      const target = await resolveTarget(context, args.targetId ? String(args.targetId) : '');
      const outputPath = resolveWithinRoot(context.config.rootDir, String(args.outputPath));
      return withTargetSession(context, target, async (session) => {
        await session.send('Page.enable');
        let metricsOverridden = false;
        const format = args.format ? String(args.format) : 'png';
        try {
          const screenshotArgs = {
            format,
            fromSurface: args.fromSurface !== false,
          };
          if (format === 'jpeg' && Number.isFinite(args.quality)) {
            screenshotArgs.quality = clampNumber(Number(args.quality), 0, 100);
          }
          if (args.fullPage) {
            const metrics = await session.send('Page.getLayoutMetrics');
            const contentSize = metrics.cssContentSize ?? metrics.contentSize ?? { width: 1280, height: 720 };
            const width = Math.max(1, Math.ceil(contentSize.width));
            const height = Math.max(1, Math.ceil(contentSize.height));
            await session.send('Emulation.setDeviceMetricsOverride', {
              mobile: false,
              width,
              height,
              deviceScaleFactor: 1,
            });
            metricsOverridden = true;
            screenshotArgs.clip = {
              x: 0,
              y: 0,
              width,
              height,
              scale: 1,
            };
          }
          const screenshot = await session.send('Page.captureScreenshot', screenshotArgs, { timeoutMs: 30000 });
          await fs.mkdir(path.dirname(outputPath), { recursive: true });
          await fs.writeFile(outputPath, Buffer.from(screenshot.data, 'base64'));
        } finally {
          if (metricsOverridden) {
            try {
              await session.send('Emulation.clearDeviceMetricsOverride');
            } catch {
              // ignore cleanup failures
            }
          }
        }
        const stat = await fs.stat(outputPath);
        return {
          target: normalizeTarget(target),
          outputPath,
          bytesWritten: stat.size,
          format,
          fullPage: Boolean(args.fullPage),
        };
      });
    },
  });
}

async function withTargetSession(context, target, callback) {
  const session = new ChromeSession(target.webSocketDebuggerUrl);
  await session.connect({ timeoutMs: 10000 });
  try {
    return await callback(session);
  } finally {
    await session.close();
  }
}

async function resolveTarget(context, requestedTargetId) {
  const targets = await listTargets(context.config);
  if (targets.length === 0) {
    throw new Error('Chrome has no available targets.');
  }
  const explicitTarget = requestedTargetId
    ? targets.find((item) => item.id === requestedTargetId)
    : null;
  let target = explicitTarget ?? null;
  if (!target && context.browserState.currentTargetId) {
    const currentTarget = targets.find((item) => item.id === context.browserState.currentTargetId);
    if (currentTarget?.type === 'page') {
      target = currentTarget;
    }
  }
  if (!target) {
    target = targets.find((item) => item.type === 'page')
      ?? targets.find((item) => item.id === context.browserState.currentTargetId)
      ?? targets[0];
  }
  context.browserState.currentTargetId = target.id;
  return target;
}

function normalizeTarget(target) {
  return {
    id: target.id,
    type: target.type,
    title: target.title ?? '',
    url: target.url ?? '',
    attached: Boolean(target.attached),
    webSocketDebuggerUrl: target.webSocketDebuggerUrl ?? '',
  };
}

function normalizeDomNode(node) {
  return {
    nodeId: node.nodeId,
    backendNodeId: node.backendNodeId,
    nodeType: node.nodeType,
    nodeName: node.nodeName,
    localName: node.localName,
    nodeValue: node.nodeValue,
    attributes: attributesToMap(node.attributes ?? []),
  };
}

function attributesToMap(attributes) {
  const map = {};
  for (let index = 0; index < attributes.length; index += 2) {
    map[attributes[index]] = attributes[index + 1];
  }
  return map;
}

function resolveWithinRoot(rootDir, value) {
  return path.isAbsolute(value) ? value : path.resolve(rootDir, value);
}

File: maya/ai_maya_agent/ai_maya_agent/tools.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .maya_runner import probe_maya_version, run_maya_operation
from .models import AiManifestBundleDefinition, AiToolDanger, AiToolDefinition, AiToolExecutionContext, clamp_number, schema_json


def default_bundles() -> list[AiManifestBundleDefinition]:
    return [
        AiManifestBundleDefinition(id="service", description="Adapter health, logs, config, and recent calls.", prefixes=["service."]),
        AiManifestBundleDefinition(id="maya.inspect", description="Maya installation, scene inspection, node inspection, and attribute inspection tools.", prefixes=["maya.installation_", "maya.scene_", "maya.nodes_", "maya.node_", "maya.attr_get"]),
        AiManifestBundleDefinition(id="maya.mutate", description="Controlled Maya node and attribute mutation tools.", prefixes=["maya.attr_set", "maya.node_create", "maya.node_delete"]),
    ]


def register_default_tools(registry, context: AiToolExecutionContext) -> None:
    def register(**kwargs: Any) -> None:
        registry.register(AiToolDefinition(**kwargs))

    register(id="service.health_get", description="Return the adapter health payload.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {}}), return_schema_json=schema_json({"type": "object"}), handler_name="service.health_get", handler=lambda args, ctx: ctx.build_health_payload(include_ok=True))
    register(id="service.agent_brief_get", description="Return the concise operating brief for this adapter.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {}}), return_schema_json=schema_json({"type": "object"}), handler_name="service.agent_brief_get", handler=lambda args, ctx: ctx.build_agent_brief_payload(include_ok=True))
    register(id="service.config_get", description="Return effective adapter configuration and current scene defaults.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {}}), return_schema_json=schema_json({"type": "object"}), handler_name="service.config_get", handler=lambda args, ctx: {"rootDir": str(ctx.config.root_dir), "requireToken": ctx.config.require_token, "fullAccessEnabled": ctx.config.full_access_enabled, "acceptedTokenHeaders": ctx.config.accepted_token_headers, "mayapyExecutable": ctx.config.mayapy_executable, "defaultSceneFile": _default_scene_file(ctx)})
    register(id="service.logs_get", description="List recent service logs.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {"maxEntries": {"type": "integer", "minimum": 1, "maximum": 300}}}), return_schema_json=schema_json({"type": "object"}), handler_name="service.logs_get", handler=lambda args, ctx: ctx.result_handle_store.build_items_result(source_tool_id="service.logs_get", field_name="logs", items=ctx.runtime_state.recent_logs(int(args.get("maxEntries") or 100)), summary={"kind": "logs"}, page_size=clamp_number(int(args.get("maxEntries") or 100), 1, 200)))
    register(id="service.recent_tool_calls_get", description="List recent tool call outcomes recorded by the adapter.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {"maxEntries": {"type": "integer", "minimum": 1, "maximum": 300}}}), return_schema_json=schema_json({"type": "object"}), handler_name="service.recent_tool_calls_get", handler=lambda args, ctx: ctx.result_handle_store.build_items_result(source_tool_id="service.recent_tool_calls_get", field_name="calls", items=ctx.runtime_state.recent_calls(int(args.get("maxEntries") or 100)), summary={"kind": "toolCalls"}, page_size=clamp_number(int(args.get("maxEntries") or 100), 1, 200)))
    register(id="service.token_regenerate", description="Regenerate the adapter token and return the new token value.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {}}), return_schema_json=schema_json({"type": "object"}), handler_name="service.token_regenerate", danger=AiToolDanger.MEDIUM, requires_confirmation=True, handler=lambda args, ctx: {"token": ctx.regenerate_token(), "acceptedTokenHeaders": ctx.config.accepted_token_headers})
    register(id="maya.installation_get", description="Probe the mayapy executable and return version details.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {}}), return_schema_json=schema_json({"type": "object"}), handler_name="maya.installation_get", handler=lambda args, ctx: probe_maya_version(ctx.config.mayapy_executable))
    register(id="maya.scene_summary_get", description="Return summary information for the selected Maya scene file or a new scene.", args_schema_json=schema_json(_scene_file_args_schema()), return_schema_json=schema_json({"type": "object"}), handler_name="maya.scene_summary_get", handler=lambda args, ctx: _run_read_operation(ctx, "scene_summary", args))
    register(id="maya.nodes_list", description="List Maya nodes with optional type filtering.", args_schema_json=schema_json({"type": "object", "additionalProperties": False, "properties": {"sceneFile": {"type": "string"}, "nodeType": {"type": "string"}, "dagOnly": {"type": "boolean"}, "longNames": {"type": "boolean"}, "scanLimit": {"type": "integer", "minimum": 1, "maximum": 20000}, "pageSize": {"type": "integer", "minimum": 1, "maximum": 200}}}), return_schema_json=schema_json({"type": "object"}), handler_name="maya.nodes_list", handler=lambda args, ctx: _paged_list_result(ctx, source_tool_id="maya.nodes_list", field_name="nodes", items=_run_read_operation(ctx, "nodes_list", args), summary={"kind": "nodes", "nodeType": args.get("nodeType", ""), "dagOnly": bool(args.get("dagOnly", False))}, page_size=int(args.get("pageSize") or 20)))
    register(id="maya.node_get", description="Return detailed metadata for one Maya node.", args_schema_json=schema_json({"type": "object", "required": ["nodeName"], "additionalProperties": False, "properties": {"sceneFile": {"type": "string"}, "nodeName": {"type": "string"}}}), return_schema_json=schema_json({"type": "object"}), handler_name="maya.node_get", handler=lambda args, ctx: _run_read_operation(ctx, "node_get", args))
    register(id="maya.attr_get", description="Return one Maya attribute value.", args_schema_json=schema_json({"type": "object", "required": ["nodeName", "attrName"], "additionalProperties": False, "properties": {"sceneFile": {"type": "string"}, "nodeName": {"type": "string"}, "attrName": {"type": "string"}}}), return_schema_json=schema_json({"type": "object"}), handler_name="maya.attr_get", handler=lambda args, ctx: _run_read_operation(ctx, "attr_get", args))
    register(id="maya.attr_set", description="Set one Maya attribute and save the result.", args_schema_json=schema_json({"type": "object", "required": ["nodeName", "attrName"], "additionalProperties": False, "properties": {"sceneFile": {"type": "string"}, "nodeName": {"type": "string"}, "attrName": {"type": "string"}, "value": {}, "values": {"type": "array"}, "outputSceneFile": {"type": "string"}, "inPlace": {"type": "boolean"}}}), return_schema_json=schema_json({"type": "object"}), handler_name="maya.attr_set", danger=AiToolDanger.HIGH, requires_confirmation=True, handler=lambda args, ctx: _run_write_operation(ctx, "attr_set", args))
    register(id="maya.node_create", description="Create a Maya node and save the result.", args_schema_json=schema_json({"type": "object", "required": ["nodeType"], "additionalProperties": False, "properties": {"sceneFile": {"type": "string"}, "nodeType": {"type": "string"}, "nodeName": {"type": "string"}, "parentNode": {"type": "string"}, "outputSceneFile": {"type": "string"}, "inPlace": {"type": "boolean"}}}), return_schema_json=schema_json({"type": "object"}), handler_name="maya.node_create", danger=AiToolDanger.HIGH, requires_confirmation=True, handler=lambda args, ctx: _run_write_operation(ctx, "node_create", args))
    register(id="maya.node_delete", description="Delete a Maya node and save the result.", args_schema_json=schema_json({"type": "object", "required": ["nodeName"], "additionalProperties": False, "properties": {"sceneFile": {"type": "string"}, "nodeName": {"type": "string"}, "outputSceneFile": {"type": "string"}, "inPlace": {"type": "boolean"}}}), return_schema_json=schema_json({"type": "object"}), handler_name="maya.node_delete", danger=AiToolDanger.HIGH, requires_confirmation=True, handler=lambda args, ctx: _run_write_operation(ctx, "node_delete", args))


def _default_scene_file(context: AiToolExecutionContext) -> str | None:
    path = _resolve_scene_file(context, {}, required=False)
    return str(path) if path else None


def _scene_file_args_schema() -> dict[str, Any]:
    return {"type": "object", "additionalProperties": False, "properties": {"sceneFile": {"type": "string"}}}


def _run_read_operation(context: AiToolExecutionContext, operation: str, args: dict[str, Any]) -> Any:
    scene_file = _resolve_scene_file(context, args, required=False)
    result = run_maya_operation(context.config, operation=operation, args=_normalize_args(context, args), scene_file=scene_file)
    return result["result"]


def _run_write_operation(context: AiToolExecutionContext, operation: str, args: dict[str, Any]) -> Any:
    scene_file = _resolve_scene_file(context, args, required=False)
    result = run_maya_operation(context.config, operation=operation, args=_normalize_args(context, args), scene_file=scene_file)
    return result["result"]


def _paged_list_result(context: AiToolExecutionContext, *, source_tool_id: str, field_name: str, items: list[Any], summary: dict[str, Any], page_size: int) -> dict[str, Any]:
    return context.result_handle_store.build_items_result(source_tool_id=source_tool_id, field_name=field_name, items=items, summary=summary, page_size=clamp_number(page_size or 20, 1, 200))


def _resolve_scene_file(context: AiToolExecutionContext, args: dict[str, Any], *, required: bool) -> Path | None:
    value = str(args.get("sceneFile") or context.config.default_scene_file or "").strip()
    if not value:
        if required:
            raise RuntimeError("This tool requires sceneFile or a configured default scene file.")
        return None
    path = _resolve_path(context.config.root_dir, value)
    if not path.exists():
        raise FileNotFoundError(f"Scene file does not exist: {path}")
    return path


def _normalize_args(context: AiToolExecutionContext, args: dict[str, Any]) -> dict[str, Any]:
    normalized = dict(args)
    if normalized.get("sceneFile"):
        normalized["sceneFile"] = str(_resolve_path(context.config.root_dir, str(normalized["sceneFile"])))
    if normalized.get("outputSceneFile"):
        normalized["outputSceneFile"] = str(_resolve_path(context.config.root_dir, str(normalized["outputSceneFile"])))
    return normalized


def _resolve_path(root_dir: Path, value: str) -> Path:
    path = Path(value).expanduser()
    return path.resolve() if path.is_absolute() else (root_dir / path).resolve()
